x/evidence/src/keeper.rs

```diff
@@ -0,0 +1,127 @@
+use crate::{
+    types::{Evidence, Handler, Router},
+    GenesisState, RouterAlreadyExistsError,
+};
+use gears::{
+    context::{init::InitContext, QueryableContext, TransactionalContext},
+    store::{
+        database::{ext::UnwrapCorrupt, Database},
+        StoreKey,
+    },
+    tendermint::informal::Hash,
+    types::store::gas::{errors::GasStoreErrors, ext::GasResultExt},
+    x::{
+        keepers::{slashing::EvidenceSlashingKeeper, staking::SlashingStakingKeeper},
+        module::Module,
+    },
+};
+use std::{collections::HashMap, marker::PhantomData};
+
+const KEY_PREFIX_EVIDENCE: [u8; 1] = [0x0];
+
+/// Keeper of the evidence store
+#[derive(Debug, Clone)]
+pub struct Keeper<
+    'a,
+    SK: StoreKey,
+    StkK: SlashingStakingKeeper<SK, M>,
+    SlsK: EvidenceSlashingKeeper<SK, M>,
+    DB: Database,
+    E: Evidence,
+    M: Module,
+> {
+    store_key: SK,
+    // TODO
+    #[allow(dead_code)]
+    staking_keeper: StkK,
+    #[allow(dead_code)]
+    slashing_keeper: SlsK,
+    router: Option<Router<'a, DB, SK, E>>,
+    _module: PhantomData<M>,
+}
+
+impl<
+        'a,
+        SK: StoreKey,
+        StkK: SlashingStakingKeeper<SK, M>,
+        SlsK: EvidenceSlashingKeeper<SK, M>,
+        DB: Database,
+        E: Evidence + Default,
+        M: Module,
+    > Keeper<'a, SK, StkK, SlsK, DB, E, M>
+{
+    pub fn new(
+        store_key: SK,
+        staking_keeper: StkK,
+        slashing_keeper: SlsK,
+        routes: Option<HashMap<String, Handler<'a, DB, SK, E>>>,
+    ) -> Self {
+        Self {
+            store_key,
+            staking_keeper,
+            slashing_keeper,
+            router: routes.map(|routes| Router::new(routes)),
+            _module: PhantomData,
+        }
+    }
+
+    /// set_router sets the Evidence Handler router for the x/evidence module. Note,
+    /// we allow the ability to set the router after the Keeper is constructed as a
+    /// given Handler may need access the Keeper before being constructed. The router
+    /// may only be set once.
+    pub fn set_router(
+        &mut self,
+        router: Router<'a, DB, SK, E>,
+    ) -> Result<(), RouterAlreadyExistsError> {
+        if self.router.is_some() {
+            Err(RouterAlreadyExistsError)
+        } else {
+            self.router = Some(router);
+            Ok(())
+        }
+    }
+
+    /// genesis initializes the evidence module's state from a provided genesis
+    /// state.
+    pub fn init_genesis(&self, ctx: &mut InitContext<'_, DB, SK>, genesis: GenesisState) {
+        if let Err(e) = genesis.validate::<E>() {
+            panic!("failed to validate evidence genesis state: {e}");
```

Comment:
What do you think? Is it worth to return errors from keeper and panic only in abci handler or it more annoying?

Reply:
Moved